### ISSUE TITLE
[UX] should we provide a UI for controlling the classes on fields?

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2994]
## Describe your issue or idea
In Backdrop core we provide a UI for selecting classes on blocks and views, but nothing else. In Drupal the Display Suite module, and the [Fences module](https://www.drupal.org/project/fences) -- and perhaps others -- provide a UI for adding classes to fields. Is this a valuable enough feature that it should be considered for core?

Display Suite UI:
![image](https://user-images.githubusercontent.com/397895/36938010-5ef1ac48-1ed0-11e8-9be4-bc68ef07542d.png)

**Comment by olafgrabienski** [comment]
Absolutely!

Reading the related issue #2993 about Multiple field settings I remembered a more general discussion about customization of the field display:

> Overriding the label text and specify customized HTML... these could be added as features in 1.x before or after any Field UI overhaul. Right now the ability to customize field display is pretty limited (just formatter settings). You're right that we want to allow for more control over markup through the UI and Field module lags behind Layout and Views modules in this regard. If we can pick out specific features and implement them individually, we can progressively iterate to add many of the DS features.
>
> https://github.com/backdrop/backdrop-issues/issues/2084#issuecomment-244277095 by @quicksketch 

As the cited comment is of an older issue with some side noise, can we expand this discussion here, so that it is not only about classes but about field display in the Field UI in general? A first stop could be to list the available field display options of Views and Block (Layout) configuration.

**Comment by jenlampton** [comment]
> A first step could be to list the available field display options of Views and Block (Layout) configuration.

Blocks simply re-use what's already in the field UI. Views does too, with the exception of multiple field settings -- hence [that issue](https://github.com/backdrop/backdrop-issues/issues/2993). 

This might qualify as a sub-issue of the [older DS issue](https://github.com/backdrop/backdrop-issues/issues/2084), but the [original Field UI  issue](https://github.com/backdrop/backdrop-issues/issues/779) is about "field display in the Field UI in general", so I don't think we should bring that discussion here.

**Comment by mazzech** [comment]
Sounds like a great feature to have this in core, cool to see it as a separate new issue. The function might help with any CSS framework with "descriptive" CSS names (e.g. div.uk-article-lead in the UiKit framework)

Perhaps another small example why I like the Dispaly Suite Field functions a lot. Heres the setting to have a Rollover Teaser available for the preview page of a certain content type.

Framework CSS: https://getuikit.com/v2/docs/overlay.html
Display Suite s
![screen shot 2018-03-04 at 07 50 29](https://user-images.githubusercontent.com/5985692/36943147-e273af7c-1f83-11e8-8c29-2132b009bb3d.png)
ettings:
![screen shot 2018-03-04 at 07 50 52](https://user-images.githubusercontent.com/5985692/36943137-c85436a2-1f83-11e8-887c-1b4803a7de3f.png)

**Comment by mazzech** [comment]
If I only had read more issues here before posting... it seems that @olafgrabienski has summed up my request some months ago in a much clearer way. We both appear to have a similar way using DS in D7... 100% support from my side what Olaf proposes in #2084. (Let’s further discuss this at DrupalCamp Ruhr later this month!

**Comment by jenlampton** [comment]
Great teamwork!

**Comment by olafgrabienski** [comment]
Yes, thanks @jenlampton for updating the issue description in #2084 !

**Comment by olafgrabienski** [comment]
> Blocks simply re-use what's already in the field UI. Views does too (...).

Not sure if I understand you correctly. The Views field display configuration for instance is much more flexible than the Content field display configuration. On a Views field, I'm able to

- create a display label and choose the name of it
-  place a colon after the label, or not
- customize the field HTML (element and classes)
- customize the label HTML (element and classes)
- customize the wrapper HTML (element and classes)

A field block in a Layout is similar. I just need to choose the *Dynamic* style and will be able to determine HTML elements and classes of wrapper, heading, and content. (AFAIK I'm not able to change the name of the label, though.)

**Comment by jenlampton** [comment]
My point was that whatever we add to the field UI will also show up in Blocks & views (since Blocks & Views uses that same code). So, if we end up adding field and field-wrapper controls to the filed UI, those will also show up in views & blocks UIs.  It will need to be done carefully so that we don't end up with duplicate options in those places when we're done...  unless we want them there too? 

A thought: This kind of change to something like fields may also have a significant impact on performance. We should do some before/after testing to be sure.

**Comment by mazzech** [comment]
@olafgrabienski regarding Views, a Drupal core developer told me years ago to *never* use the field mode in Vies, but going for the Entity mode instead whenever possible (much better performance). Since then I started to do so in 90% of the cases, and found it to be a much better solution over all. As far as I understand, having a Entity based solution would also help for non-Views scenarios, but also I need to further use the Backdrop layout system to make sure...

**Comment by olafgrabienski** [comment]
@mazzech Thanks for the tip! I agree partially but I've continued to use views fields a lot. You can just do so much with them, and fortunately I've had rarely performance issues so far. So, I use Display suite  mainly for full content display and other display modes (e.g. referenced content).

**Comment by klonos** [comment]
Should we consolidate/merge this with #2608? (although this question/request here is only for classes, whereas that other issue is about wrapper HTML classes too)

**Comment by olafgrabienski** [comment]
> Should we consolidate/merge this with https://github.com/backdrop/backdrop-issues/issues/2608?

If you're asking me, yes! As claimed in the [last comment of the other issue](https://github.com/backdrop/backdrop-issues/issues/2608#issuecomment-504351173), in my opinion three related features "would meet the basic needs of many site builders":
>
> - Change the HTML elements of field item(s) and wrapper (Fences, or [Field Wrappers](https://github.com/backdrop-contrib/field_wrappers))
> - Add classes to the HTML elements and the label (Fences)
> - Change the name of the label.

**Comment by klonos** [comment]
@jenlampton since you opened this issue, do you agree to close it in favor of #2608?

**Comment by olafgrabienski** [comment]
@jenlampton Friendly reminder:

> do you agree to close [this issue] in favor of https://github.com/backdrop/backdrop-issues/issues/2608?
